File: javascript/javascript/Python/JavaScriptLexerBase.py

from antlr4 import *

relativeImport = False
if __name__ is not None and "." in __name__:
    relativeImport = True

class JavaScriptLexerBase(Lexer):
    def __init__(self, *args, **kwargs):
        super(JavaScriptLexerBase, self).__init__(*args, **kwargs)

        """Stores values of nested modes. By default mode is strict or
        defined externally (useStrictDefault)"""
        self.scopeStrictModes = []
        self.lastToken: Token = None

        """Default value of strict mode
        Can be defined externally by setUseStrictDefault"""
        self.useStrictDefault = False

        """Current value of strict mode
        Can be defined during parsing, see StringFunctions.js and StringGlobal.js samples"""
        self.useStrictCurrent = False

    def getStrictDefault(self) -> bool:
        return self.useStrictDefault

    def setUseStrictDefault(self, value: bool):
        self.useStrictDefault = value
        self.useStrictCurrent = value

    def isStrictMode(self):
        return self.useStrictCurrent

    def isStartOfFile(self):
        return self.lastToken is None

    def nextToken(self) -> Token:
        """Return the next token from the character stream and records this last
        token in case it resides on the default channel. This recorded token
        is used to determine when the lexer could possibly match a regex
        literal. Also changes scopeStrictModes stack if tokenize special
        string 'use strict';

        :return the next token from the character stream."""
        next_token: Token = super(JavaScriptLexerBase, self).nextToken()

        if next_token.channel == Token.DEFAULT_CHANNEL:
            self.lastToken = next_token

        return next_token

    def processOpenBrace(self):
        self.useStrictCurrent = bool(self.scopeStrictModes) and (True if self.scopeStrictModes[-1] else self.useStrictDefault)
        self.scopeStrictModes.append(self.useStrictCurrent)

    def processCloseBrace(self):
        self.useStrictCurrent = bool(self.scopeStrictModes) and (True if self.scopeStrictModes.pop(-1) else self.useStrictDefault)

    def processStringLiteral(self):
        if relativeImport:
            from .JavaScriptLexer import JavaScriptLexer
        else:
            from JavaScriptLexer import JavaScriptLexer
        if not self.lastToken or self.lastToken.type == JavaScriptLexer.OpenBrace:
            text = self.text
            if text == '"use strict"' or text == "'use strict'":
                if self.scopeStrictModes:
                    self.scopeStrictModes.pop(-1)
                self.useStrictCurrent = True
                self.scopeStrictModes.append(self.useStrictCurrent)

    def isRegexPossible(self) -> bool:
        """Returns {@code true} if the lexer can match a regex literal. """
        if relativeImport:
            from .JavaScriptLexer import JavaScriptLexer
        else:
            from JavaScriptLexer import JavaScriptLexer

        if not self.lastToken:
            # No token has been produced yet: at the start of the input,
            # no division is possible, so a regex literal _is_ possible.
            return True

        if self.lastToken.type in [
                JavaScriptLexer.Identifier,
                JavaScriptLexer.NullLiteral,
                JavaScriptLexer.BooleanLiteral,
                JavaScriptLexer.This,
                JavaScriptLexer.CloseBracket,
                JavaScriptLexer.CloseParen,
                JavaScriptLexer.OctalIntegerLiteral,
                JavaScriptLexer.DecimalLiteral,
                JavaScriptLexer.HexIntegerLiteral,
                JavaScriptLexer.StringLiteral,
                JavaScriptLexer.PlusPlus,
                JavaScriptLexer.MinusMinus]:
            return False

        return True

File: javascript/javascript/Java/JavaScriptLexerBase.java
import org.antlr.v4.runtime.*;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * All lexer methods that used in grammar (IsStrictMode)
 * should start with Upper Case Char similar to Lexer rules.
 */
public abstract class JavaScriptLexerBase extends Lexer
{
    /**
     * Stores values of nested modes. By default mode is strict or
     * defined externally (useStrictDefault)
     */
    private final Deque<Boolean> scopeStrictModes = new ArrayDeque<>();

    private Token lastToken = null;
    /**
     * Default value of strict mode
     * Can be defined externally by setUseStrictDefault
     */
    private boolean useStrictDefault = false;
    /**
     * Current value of strict mode
     * Can be defined during parsing, see StringFunctions.js and StringGlobal.js samples
     */
    private boolean useStrictCurrent = false;
    /**
     * Preserves depth due to braces including template literals.
     */
	private int currentDepth = 0;

    /**
     * Preserves the starting depth of template literals to correctly handle braces inside template literals.
     */
	private Deque<Integer> templateDepthStack = new ArrayDeque<Integer>();

    public JavaScriptLexerBase(CharStream input) {
        super(input);
    }

    public boolean IsStartOfFile() {
        return lastToken == null;
    }

    public boolean getStrictDefault() {
        return useStrictDefault;
    }

    public void setUseStrictDefault(boolean value) {
        useStrictDefault = value;
        useStrictCurrent = value;
    }

    public boolean IsStrictMode() {
        return useStrictCurrent;
    }

    public boolean IsInTemplateString() {
		return !templateDepthStack.isEmpty() && templateDepthStack.peek() == currentDepth;
    }

    /**
     * Return the next token from the character stream and records this last
     * token in case it resides on the default channel. This recorded token
     * is used to determine when the lexer could possibly match a regex
     * literal. Also changes scopeStrictModes stack if tokenize special
     * string 'use strict';
     *
     * @return the next token from the character stream.
     */
    @Override
    public Token nextToken() {
        Token next = super.nextToken();

        if (next.getChannel() == Token.DEFAULT_CHANNEL) {
            // Keep track of the last token on the default channel.
            this.lastToken = next;
        }

        return next;
    }

    protected void ProcessOpenBrace()
    {
		currentDepth++;
        useStrictCurrent = scopeStrictModes.size() > 0 && scopeStrictModes.peek() ? true : useStrictDefault;
        scopeStrictModes.push(useStrictCurrent);
    }

    protected void ProcessCloseBrace()
    {
        useStrictCurrent = scopeStrictModes.size() > 0 ? scopeStrictModes.pop() : useStrictDefault;
		currentDepth--;
    }

	protected void ProcessTemplateOpenBrace() {
		currentDepth++;
		this.templateDepthStack.push(currentDepth);
	}

	protected void ProcessTemplateCloseBrace() {
		this.templateDepthStack.pop();
		currentDepth--;
	}

    protected void ProcessStringLiteral()
    {
        if (lastToken == null || lastToken.getType() == JavaScriptLexer.OpenBrace)
        {
            String text = getText();
            if (text.equals("\"use strict\"") || text.equals("'use strict'"))
            {
                if (scopeStrictModes.size() > 0)
                    scopeStrictModes.pop();
                useStrictCurrent = true;
                scopeStrictModes.push(useStrictCurrent);
            }
        }
    }

    /**
     * Returns {@code true} if the lexer can match a regex literal.
     */
    protected boolean IsRegexPossible() {

        if (this.lastToken == null) {
            // No token has been produced yet: at the start of the input,
            // no division is possible, so a regex literal _is_ possible.
            return true;
        }

        switch (this.lastToken.getType()) {
            case JavaScriptLexer.Identifier:
            case JavaScriptLexer.NullLiteral:
            case JavaScriptLexer.BooleanLiteral:
            case JavaScriptLexer.This:
            case JavaScriptLexer.CloseBracket:
            case JavaScriptLexer.CloseParen:
            case JavaScriptLexer.OctalIntegerLiteral:
            case JavaScriptLexer.DecimalLiteral:
            case JavaScriptLexer.HexIntegerLiteral:
            case JavaScriptLexer.StringLiteral:
            case JavaScriptLexer.PlusPlus:
            case JavaScriptLexer.MinusMinus:
                // After any of the tokens above, no regex literal can follow.
                return false;
            default:
                // In all other cases, a regex literal _is_ possible.
                return true;
        }
    }

    @Override
    public void reset() {
        this.scopeStrictModes.clear();
        this.lastToken = null;
        this.useStrictDefault = false;
        this.useStrictCurrent = false;
	    this.currentDepth = 0;
	    this.templateDepthStack = new ArrayDeque<Integer>();
        super.reset();
    }
}
